rollup/core/classes/RollupCalculator.cls

```diff
@@ -1348,19 +1345,103 @@ public without sharing abstract class RollupCalculator {
         this.fieldNames.add(trimmedFieldName);
       }
     }
+  }
+
+  private interface GroupingFormatter {
+    String getGroupingStart();
+    String getGroupingColumn(String fieldName, Object groupingValue, Boolean isFirst, Boolean isLast);
+    String getGroupingRowStart(String groupingKey);
+    String getGroupingRowEnd();
+    String getEnd(String groupByString);
+  }
+
+  private class DefaultGroupingFormatter implements GroupingFormatter {
+    private final String delimiter;
+    private final Rollup__mdt metadata;
+
+    public DefaultGroupingFormatter(Rollup__mdt metadata, String delimiter, Schema.SObjectField opFieldOnLookupObject) {
+      this.metadata = metadata;
+      this.delimiter = delimiter;
+      this.setupDelimiters(opFieldOnLookupObject);
+    }
+
+    public String getGroupingColumn(String fieldName, Object groupingValue, Boolean isFirst, Boolean isLast) {
+      return String.valueOf(groupingValue) + this.delimiter;
+    }
+
+    public String getGroupingStart() {
+      return '';
+    }
+
+    public String getGroupingRowStart(String groupingKey) {
+      return this.metadata.GroupByRowStartDelimiter__c + groupingKey;
+    }
+
+    public String getGroupingRowEnd() {
+      return this.metadata.GroupByRowEndDelimiter__c;
+    }
+
+    public String getEnd(String groupByString) {
+      return groupByString.removeEnd(this.getGroupingRowEnd());
+    }
 
-    private void setupDelimiters() {
+    private void setupDelimiters(Schema.SObjectField opFieldOnLookupObject) {
       if (this.metadata.GroupByRowEndDelimiter__c == null) {
         // insane hack to identify rich text fields
-        Boolean isRichText = JSON.serialize(this.opFieldOnLookupObject.getDescribe()).contains('"extraTypeInfo":"richtextarea"');
+        Boolean isRichText = JSON.serialize(opFieldOnLookupObject.getDescribe()).contains('"extraTypeInfo":"richtextarea"');
         this.metadata.GroupByRowEndDelimiter__c = isRichText ? '<br>' : '\n';
       }
+
       if (this.metadata.GroupByRowStartDelimiter__c == null) {
         this.metadata.GroupByRowStartDelimiter__c = '• ';
       }
     }
   }
 
+  private class TableGroupingFormatter implements GroupingFormatter {
+    private final Map<String, Schema.SObjectField> fieldNameToToken;
+    private final Schema.SObjectField targetField;
+    private final Set<String> columnHeaders = new Set<String>();
+
+    private final String tdOpen = '<td>';
+    private final String tdClose = '</td>';
+    private final String thOpen = '<th>';
+    private final String thClose = '</th>';
+    private final String trOpen = '<tr>';
+    private final String trClose = '</tr>';
```

Comment:
@jongpie I've gone back and forth on this section. It eliminates some duplication at the cost of readability. This class isn't perfect, by any means, so I'll be curious to hear what you think! In particular, I knowingly skipped over adding tags like `<thead>` and `<tbody>` as the complexity of the simple version of the table evolved to ... this

Reply:
Considering you're generating HTML inside of Apex, I don't hate this approach - the class `TableGroupingFormatter` feels pretty clean & readable to me. And yeah, for this purpose, if excluding `<thead>` and `<tbody>` doesn't impact how the table is rendered, then I say don't worry about them!

Reply:
Thank you for re-reviewing! I really appreciate it.